### PR TITLE
r/cloudwatch_metric_alarm - add validations

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #13624
Relates #13826

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource_aws_cloudwatch_metric_alarm: add plan time validations to `comparison_operator` and `statistic`, `metric_query.metric.unit`, `unit`.
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSCloudWatchMetricAlarm_'
--- PASS: TestAccAWSCloudWatchMetricAlarm_missingStatistic (14.22s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_disappears (32.10s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_datapointsToAlarm (36.33s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_extendedStatistic (38.08s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_basic (43.27s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_AlarmActions_SWFAction (49.03s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_AlarmActions_SNSTopic (50.49s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_treatMissingData (70.35s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_evaluateLowSampleCountPercentiles (76.07s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_tags (107.46s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_expression (168.96s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_AlarmActions_EC2Automate (340.33s)
```

### CODE REVIEW

**Comment by ewbankkit** [comment]
Verified acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSCloudWatchMetricAlarm_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAWSCloudWatchMetricAlarm_ -timeout 120m
=== RUN   TestAccAWSCloudWatchMetricAlarm_basic
=== PAUSE TestAccAWSCloudWatchMetricAlarm_basic
=== RUN   TestAccAWSCloudWatchMetricAlarm_AlarmActions_EC2Automate
=== PAUSE TestAccAWSCloudWatchMetricAlarm_AlarmActions_EC2Automate
=== RUN   TestAccAWSCloudWatchMetricAlarm_AlarmActions_SNSTopic
=== PAUSE TestAccAWSCloudWatchMetricAlarm_AlarmActions_SNSTopic
=== RUN   TestAccAWSCloudWatchMetricAlarm_AlarmActions_SWFAction
=== PAUSE TestAccAWSCloudWatchMetricAlarm_AlarmActions_SWFAction
=== RUN   TestAccAWSCloudWatchMetricAlarm_datapointsToAlarm
=== PAUSE TestAccAWSCloudWatchMetricAlarm_datapointsToAlarm
=== RUN   TestAccAWSCloudWatchMetricAlarm_treatMissingData
=== PAUSE TestAccAWSCloudWatchMetricAlarm_treatMissingData
=== RUN   TestAccAWSCloudWatchMetricAlarm_evaluateLowSampleCountPercentiles
=== PAUSE TestAccAWSCloudWatchMetricAlarm_evaluateLowSampleCountPercentiles
=== RUN   TestAccAWSCloudWatchMetricAlarm_extendedStatistic
=== PAUSE TestAccAWSCloudWatchMetricAlarm_extendedStatistic
=== RUN   TestAccAWSCloudWatchMetricAlarm_expression
=== PAUSE TestAccAWSCloudWatchMetricAlarm_expression
=== RUN   TestAccAWSCloudWatchMetricAlarm_missingStatistic
=== PAUSE TestAccAWSCloudWatchMetricAlarm_missingStatistic
=== RUN   TestAccAWSCloudWatchMetricAlarm_tags
=== PAUSE TestAccAWSCloudWatchMetricAlarm_tags
=== RUN   TestAccAWSCloudWatchMetricAlarm_disappears
=== PAUSE TestAccAWSCloudWatchMetricAlarm_disappears
=== CONT  TestAccAWSCloudWatchMetricAlarm_AlarmActions_EC2Automate
=== CONT  TestAccAWSCloudWatchMetricAlarm_disappears
=== CONT  TestAccAWSCloudWatchMetricAlarm_tags
=== CONT  TestAccAWSCloudWatchMetricAlarm_missingStatistic
=== CONT  TestAccAWSCloudWatchMetricAlarm_expression
=== CONT  TestAccAWSCloudWatchMetricAlarm_extendedStatistic
=== CONT  TestAccAWSCloudWatchMetricAlarm_evaluateLowSampleCountPercentiles
=== CONT  TestAccAWSCloudWatchMetricAlarm_treatMissingData
=== CONT  TestAccAWSCloudWatchMetricAlarm_datapointsToAlarm
=== CONT  TestAccAWSCloudWatchMetricAlarm_AlarmActions_SWFAction
=== CONT  TestAccAWSCloudWatchMetricAlarm_AlarmActions_SNSTopic
=== CONT  TestAccAWSCloudWatchMetricAlarm_basic
--- PASS: TestAccAWSCloudWatchMetricAlarm_missingStatistic (11.28s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_disappears (21.22s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_datapointsToAlarm (24.66s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_extendedStatistic (25.74s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_basic (28.96s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_AlarmActions_SWFAction (31.30s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_AlarmActions_SNSTopic (32.19s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_evaluateLowSampleCountPercentiles (42.42s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_treatMissingData (42.51s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_tags (63.80s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_expression (101.23s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_AlarmActions_EC2Automate (228.22s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	228.265s
```

**Comment by ewbankkit** [comment]
Verified acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSCloudWatchMetricAlarm_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAWSCloudWatchMetricAlarm_ -timeout 120m
go: downloading github.com/aws/aws-sdk-go v1.31.0
go: extracting github.com/aws/aws-sdk-go v1.31.0
go: finding github.com/aws/aws-sdk-go v1.31.0
=== RUN   TestAccAWSCloudWatchMetricAlarm_basic
=== PAUSE TestAccAWSCloudWatchMetricAlarm_basic
=== RUN   TestAccAWSCloudWatchMetricAlarm_AlarmActions_EC2Automate
=== PAUSE TestAccAWSCloudWatchMetricAlarm_AlarmActions_EC2Automate
=== RUN   TestAccAWSCloudWatchMetricAlarm_AlarmActions_SNSTopic
=== PAUSE TestAccAWSCloudWatchMetricAlarm_AlarmActions_SNSTopic
=== RUN   TestAccAWSCloudWatchMetricAlarm_AlarmActions_SWFAction
=== PAUSE TestAccAWSCloudWatchMetricAlarm_AlarmActions_SWFAction
=== RUN   TestAccAWSCloudWatchMetricAlarm_datapointsToAlarm
=== PAUSE TestAccAWSCloudWatchMetricAlarm_datapointsToAlarm
=== RUN   TestAccAWSCloudWatchMetricAlarm_treatMissingData
=== PAUSE TestAccAWSCloudWatchMetricAlarm_treatMissingData
=== RUN   TestAccAWSCloudWatchMetricAlarm_evaluateLowSampleCountPercentiles
=== PAUSE TestAccAWSCloudWatchMetricAlarm_evaluateLowSampleCountPercentiles
=== RUN   TestAccAWSCloudWatchMetricAlarm_extendedStatistic
=== PAUSE TestAccAWSCloudWatchMetricAlarm_extendedStatistic
=== RUN   TestAccAWSCloudWatchMetricAlarm_expression
=== PAUSE TestAccAWSCloudWatchMetricAlarm_expression
=== RUN   TestAccAWSCloudWatchMetricAlarm_missingStatistic
=== PAUSE TestAccAWSCloudWatchMetricAlarm_missingStatistic
=== RUN   TestAccAWSCloudWatchMetricAlarm_tags
=== PAUSE TestAccAWSCloudWatchMetricAlarm_tags
=== RUN   TestAccAWSCloudWatchMetricAlarm_disappears
=== PAUSE TestAccAWSCloudWatchMetricAlarm_disappears
=== CONT  TestAccAWSCloudWatchMetricAlarm_basic
=== CONT  TestAccAWSCloudWatchMetricAlarm_extendedStatistic
=== CONT  TestAccAWSCloudWatchMetricAlarm_disappears
=== CONT  TestAccAWSCloudWatchMetricAlarm_tags
=== CONT  TestAccAWSCloudWatchMetricAlarm_missingStatistic
=== CONT  TestAccAWSCloudWatchMetricAlarm_expression
=== CONT  TestAccAWSCloudWatchMetricAlarm_AlarmActions_SWFAction
=== CONT  TestAccAWSCloudWatchMetricAlarm_evaluateLowSampleCountPercentiles
=== CONT  TestAccAWSCloudWatchMetricAlarm_treatMissingData
=== CONT  TestAccAWSCloudWatchMetricAlarm_datapointsToAlarm
=== CONT  TestAccAWSCloudWatchMetricAlarm_AlarmActions_SNSTopic
=== CONT  TestAccAWSCloudWatchMetricAlarm_AlarmActions_EC2Automate
--- PASS: TestAccAWSCloudWatchMetricAlarm_missingStatistic (10.93s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_disappears (21.02s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_datapointsToAlarm (23.25s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_extendedStatistic (24.78s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_basic (27.06s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_AlarmActions_SWFAction (29.07s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_AlarmActions_SNSTopic (30.23s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_evaluateLowSampleCountPercentiles (40.76s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_treatMissingData (40.92s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_tags (60.03s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_expression (96.70s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_AlarmActions_EC2Automate (213.34s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	213.386s
```

File: .changelog/12817.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+```release-note:enhancement
+resource/aws_cloudwatch_metric_alarm: Add plan time validation to `alarm_name`, 
+`comparison_operator`, `metric_name`, `metric_query.id`, `metric_query.expression`, 
+`metric_query.metric.metric_name`, `metric_query.metric.namespace`, `metric_query.metric.unit`,
+`namespace`, `period`, `statistic`, `alarm_description`, `insufficient_data_actions`, `ok_actions`, `unit`, and `extended_statistic`
```
Comment (bflad) [inline .changelog/12817.txt:5]: Newlines should be omitted from changelog entries. Reference: https://github.com/hashicorp/terraform-provider-aws/issues/17135

```suggestion
resource/aws_cloudwatch_metric_alarm: Add plan time validation to `alarm_name`, `comparison_operator`, `metric_name`, `metric_query.id`, `metric_query.expression`, `metric_query.metric.metric_name`, `metric_query.metric.namespace`, `metric_query.metric.unit`, `namespace`, `period`, `statistic`, `alarm_description`, `insufficient_data_actions`, `ok_actions`, `unit`, and `extended_statistic`
```

File: aws/resource_aws_cloudwatch_metric_alarm.go [hunk 2]
```diff
@@ -75,12 +81,17 @@ func resourceAwsCloudWatchMetricAlarm() *schema.Resource {
...
+										Type:         schema.TypeString,
+										Required:     true,
+										ValidateFunc: validation.StringLenBetween(1, 255),
 									},
 									"namespace": {
 										Type:     schema.TypeString,
 										Optional: true,
+										ValidateFunc: validation.Any(
+											validation.StringLenBetween(1, 255),
+											validation.StringMatch(regexp.MustCompile(`[^:].*`), ""),
```
Comment (bflad) [inline aws/resource_aws_cloudwatch_metric_alarm.go:93]:
```suggestion
											validation.StringMatch(regexp.MustCompile(`[^:].*`), "must not contain colon characters"),
```

File: aws/resource_aws_cloudwatch_metric_alarm.go [hunk 3]
```diff
@@ -113,16 +125,25 @@ func resourceAwsCloudWatchMetricAlarm() *schema.Resource {
 				Type:          schema.TypeString,
 				Optional:      true,
 				ConflictsWith: []string{"metric_query"},
+				ValidateFunc: validation.Any(
+					validation.StringLenBetween(1, 255),
+					validation.StringMatch(regexp.MustCompile(`[^:].*`), ""),
```
Comment (bflad) [inline aws/resource_aws_cloudwatch_metric_alarm.go:130]:
```suggestion
					validation.StringMatch(regexp.MustCompile(`[^:].*`), "must not contain colon characters"),
```

**Review by bflad — APPROVED** [review]
Looks good to me, thank you, @DrFaust92 🚀 

References:

- https://docs.aws.amazon.com/sdk-for-go/api/service/cloudwatch/#pkg-constants
- https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_PutMetricAlarm.html

Output from acceptance testing in AWS Commercial:

```
--- PASS: TestAccAWSCloudWatchMetricAlarm_AlarmActions_EC2Automate (177.84s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_AlarmActions_SNSTopic (26.40s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_AlarmActions_SWFAction (25.53s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_basic (25.28s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_datapointsToAlarm (22.65s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_disappears (19.51s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_evaluateLowSampleCountPercentiles (33.19s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_expression (67.61s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_extendedStatistic (22.59s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_missingStatistic (11.62s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_tags (47.01s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_treatMissingData (33.18s)
```

Output from acceptance testing in AWS GovCloud (US):

```
--- PASS: TestAccAWSCloudWatchMetricAlarm_AlarmActions_EC2Automate (199.57s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_AlarmActions_SNSTopic (23.82s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_AlarmActions_SWFAction (23.14s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_basic (22.45s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_datapointsToAlarm (19.66s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_disappears (16.42s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_evaluateLowSampleCountPercentiles (33.05s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_expression (78.23s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_extendedStatistic (18.22s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_missingStatistic (8.10s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_tags (48.94s)
--- PASS: TestAccAWSCloudWatchMetricAlarm_treatMissingData (33.07s)
```

**Pull request merged**